src/transformers/tokenization_utils_base.py

```diff
@@ -1637,9 +1637,11 @@ def convert_added_tokens(obj: Union[AddedToken, Any]):
         if special_tokens_map_file is not None:
             with open(special_tokens_map_file, encoding="utf-8") as special_tokens_map_handle:
                 special_tokens_map = json.load(special_tokens_map_handle)
-
-            special_tokens_map = convert_added_tokens(special_tokens_map)
             for key, value in special_tokens_map.items():
+                if isinstance(value, dict):
+                    value = AddedToken(**value)
+                elif isinstance(value, list):
+                    value = [AddedToken(**token) if isinstance(token, dict) else token for token in value]
```

Comment:
The difference between the old (and proposed here) method and the new method (leveraging `convert_added_token`) is that the one leveraging `convert_added_token` manages an unlimited level of nesting as it's recursive, @thomwolf?

Isn't it possible to still leverage the `convert_added_token` method?

Reply:
I could write a new recursive `convert_added_token` function. But the one used previously does not seem to work for RAG. The line: `if isinstance(obj, dict) and "__type" in obj and obj["__type"] == "AddedToken":` does not seem to fit for RAG's tokenizer `facebook/rag-token-nq`. 

Also not sure if the `special_token_map` would ever have more than 2 levels of nesting, but @LysandreJik @thomwolf @n1t0 probably know better here!

Reply:
I see indeed. We need to update the `save_pretrained` method as well then.

The previous behavior of considering all dict objects as `AddedToken` seemed a bit fragile to me.
I'll update this a little bit if you don't mind, probably we can just check that all the keys in the dict are in the fields of the AddedToken class before converting it.